From 0ad696e58925fd6deba8c47a8023a7364c080594 Mon Sep 17 00:00:00 2001
From: mukawa <mukawa@igel.co.jp>
Date: Wed, 15 Aug 2012 10:41:59 +0900
Subject: [PATCH 78/80] video: rcarfb: Remove busy loops for vsync

This patch removes busy loops to monitor the vertical blanking flag.

Signed-off-by: Katsuya Matsubara <matsu@igel.co.jp>
---
 drivers/video/rcarfb.c |   26 ++++++++++++++++++--------
 1 files changed, 18 insertions(+), 8 deletions(-)

diff --git a/drivers/video/rcarfb.c b/drivers/video/rcarfb.c
index 5b7562a..605fd04 100644
--- a/drivers/video/rcarfb.c
+++ b/drivers/video/rcarfb.c
@@ -94,6 +94,9 @@ struct rcar_du_priv {
 	struct clk	*duclk;
 	int		clock_info;
 	struct rcar_reso_info *dispdev;
+
+	wait_queue_head_t wait_vbk2;
+	unsigned char vbk2;
 };
 
 static struct fb_fix_screeninfo rcar_du_fix  = {
@@ -482,6 +485,12 @@ static irqreturn_t rcar_du_irq(int irq, void *data)
 	du_write(priv, DSRCR, intr1);
 	du_write(priv, D2SRCR, intr2);
 
+	/* wake up */
+	if ((intr2 & VBK2_D2SSR) == VBK2_D2SSR) {
+		priv->vbk2 = 1;
+		wake_up(&priv->wait_vbk2);
+	}
+
 	return IRQ_HANDLED;
 }
 
@@ -707,10 +716,10 @@ static int du_update_displaystart(int fb_num, struct rcar_du_priv *priv, CarE1Pl
 	tmp = du_read(priv, PnMR(index1));	// FIXME: only for one plane?
 	du_write(priv, PnMR(index1), PnDC_ON_PnMR | tmp);
 
-	while (1) {
-		if ((du_read(priv, D2SSR) & VBK2_D2SSR) == VBK2_D2SSR)
-			break;
-	}
+	/* wait vbk2 */
+	priv->vbk2 = 0;
+	wait_event_interruptible(priv->wait_vbk2,
+		priv->vbk2 == 1);
 
 	return 0;
 }
@@ -742,10 +751,10 @@ static int rcar_du_pan_display(struct fb_var_screeninfo *var,
 		du_write(priv, PnMR(index1), PnDC_ON_PnMR | tmp);
 	}
 
-	while (1) {
-		if ((du_read(priv, D2SSR) & VBK2_D2SSR) == VBK2_D2SSR)
-			break;
-	}
+	/* wait vbk2 */
+	priv->vbk2 = 0;
+	wait_event_interruptible(priv->wait_vbk2,
+		priv->vbk2 == 1);
 
 	return 0;
 }
@@ -984,6 +993,7 @@ static int __devinit rcar_du_probe(struct platform_device *pdev)
 
 	priv->irq = i;
 	priv->base = ioremap_nocache(res->start, (res->end - res->start) + 1);
+	init_waitqueue_head(&priv->wait_vbk2);
 
 	for (i = 0; i < RESOLUTION_NUM; i++) {
 		priv->dispdev = reso_info = pdev->dev.platform_data;
-- 
1.7.0.4

